crypto/types/compact_bit_array.go

```diff
@@ -75,13 +75,22 @@ func (bA *CompactBitArray) SetIndex(i int, v bool) bool {
 // there are two bits set to true before index 4.
 func (bA *CompactBitArray) NumTrueBitsBefore(index int) int {
 	numTrueValues := 0
-	for i := 0; i < index; i++ {
-		if bA.GetIndex(i) {
-			numTrueValues++
+	max := bA.Count()
+	if index > max {
+		index = max
+	}
+	var i = 0
+	for elem := 0; ; elem++ {
```

Comment:
Can you add a comment here explaining whats happening, now that its not wrapped in ba.GetIndex(i)

Reply:
What do  you mean? The function is working any state. As you see we limit the traversing to the `max` - which is the _current_ array length - so it works with future reads.

Reply:
No, I meant future readers of the code haha. (Like just adding a code comment to the function for how it works)